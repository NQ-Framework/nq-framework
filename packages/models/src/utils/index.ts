import { PropertyValue } from "../workflow/property/property-value";

export const reducePropertyValuesToObject = (values: PropertyValue[]) => {
  return values.reduce((obj: any, prop: PropertyValue) => {
    obj[prop.name] = prop.value;
    return obj;
  }, {});
};
