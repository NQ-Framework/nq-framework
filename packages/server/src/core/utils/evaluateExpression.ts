import { NodeVM, NodeVMOptions } from 'vm2';
import {format, addDays, subDays, addHours, subHours }from "date-fns";

export const evaluateExpression = async (
  expression: string,
  context: any,
): Promise<any> => {
  const sandbox = {
    ...context,
    dateFns: {format, addDays, subDays, addHours, subHours}
  };

  const options: NodeVMOptions = {
    console: 'inherit',
    sandbox,
    require: {
      external: false as boolean,
      builtin: [] as string[],
    },
  };
  const vm = new NodeVM(options);
  const result = await vm.run(
    `module.exports = async function() {${expression}}()`,
    __dirname,
  );
  return result;
};
