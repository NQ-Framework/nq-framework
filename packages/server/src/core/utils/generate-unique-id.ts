import { getFirebaseApp } from "../../firebase/initialize"
export const generateUniqueId = async (): Promise<string> => {
    const app = await getFirebaseApp()
    return app.firestore().collection("asd").doc().id;
}