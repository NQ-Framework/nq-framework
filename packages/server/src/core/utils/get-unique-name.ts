export const getUniqueName = (desiredName: string, takenNames: string[]): string => {
    let index = 0;
    while (takenNames.includes(desiredName + (index > 0 ? " " + index : ""))) {
        index++;
    }
    return desiredName + (index > 0 ? " " + index : "");
}