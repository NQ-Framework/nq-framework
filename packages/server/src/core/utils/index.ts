export { evaluateExpression } from './evaluateExpression';
export { evaluateProperties } from './evaluateProperties';
export { getUniqueName } from './get-unique-name';
export { generateUniqueId } from './generate-unique-id';
