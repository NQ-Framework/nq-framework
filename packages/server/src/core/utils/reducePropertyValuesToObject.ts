import { PropertyValue } from '@nqframework/models';

export const reducePropertyValuesToObject = (values: PropertyValue[]) => {
  return values.reduce((obj: any, prop: PropertyValue) => {
    obj[prop.name] = prop.value;
    return obj;
  }, {});
};
