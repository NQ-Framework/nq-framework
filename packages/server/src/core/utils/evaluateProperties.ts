import { PropertyValue } from '@nqframework/models';
import { evaluateExpression } from './evaluateExpression';

export const evaluateProperties = async (
  properties: PropertyValue[],
  context: any,
): Promise<PropertyValue[]> => {
  if (!properties || !Boolean(properties.length)) {
    return [];
  }
  const propertyValues: PropertyValue[] = [];
  for (let i = 0; i < properties.length; i++) {
    const property = properties[i];
    if (property.value === undefined || property.value === '') {
      continue;
    }
    if (typeof property.value === 'string' && property.value.startsWith('=')) {
      const result = await evaluateExpression(
        `return ${property.value.slice(1)}`,
        context,
      );
      propertyValues.push({ name: property.name, value: result });
    }
    else if (Array.isArray(property.value)) {
       const result = await evaluateProperties(property.value, context);
      propertyValues.push({ name: property.name, value: result });
    }
    else {
      propertyValues.push({
        name: property.name,
        value: property.value,
      });
    }
  }
  return propertyValues;
};
