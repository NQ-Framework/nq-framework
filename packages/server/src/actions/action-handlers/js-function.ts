import {
  ActionHandler,
  ActionInstance,
  PropertyValue,
  WorkflowExecutionContext,
} from '@nqframework/models';
import { evaluateExpression } from '../../core/utils';

export const handler: ActionHandler = {
  handle: async (
    propertyValues: PropertyValue[],
    actionInstance: ActionInstance,
    workflowExecution: WorkflowExecutionContext,
  ): Promise<PropertyValue[]> => {
    const code = propertyValues.find((i) => i.name === 'code')?.value;
    if (!code) {
      throw new Error('Missing required parameter code');
    }

    const output = await evaluateExpression(code, {
      actionInstance,
      propertyValues,
      workflowExecution
    });

    return [
      {
        name: 'output',
        value: output,
      },
    ];
  },
};
