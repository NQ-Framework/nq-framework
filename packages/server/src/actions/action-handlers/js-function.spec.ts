import { handler } from './js-function';
import { evaluateExpression } from '../../core/utils/evaluateExpression';
jest.mock('../../core/utils/evaluateExpression');

describe('js function action', () => {
  it('should call vm to evaluate  code and pass the value back out', async () => {
    const mock = evaluateExpression as jest.Mock;
    mock.mockImplementation(() => {
      return 'test';
    });
    const result = await handler.handle(
      [{ name: 'code', value: 'test' }],
      {} as any,
      {} as any,
      {} as any,
    );
    expect(result).toEqual([{ name: 'output', value: 'test' }]);
    expect(mock).toHaveBeenCalledWith('test', {
      actionInstance: {},
      propertyValues: [{ name: 'code', value: 'test' }],
      workflowExecution: {},
    });
  });
  it('should throw when message is not present', async () => {
    await expect(
      handler.handle(
        [{ name: 'not code', value: 'test' }],
        {} as any,
        {} as any,
        {} as any,
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
