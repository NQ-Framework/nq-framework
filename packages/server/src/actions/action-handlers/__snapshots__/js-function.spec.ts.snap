// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`js function action should throw when message is not present 1`] = `"Missing required parameter code"`;
